### ISSUE TITLE
QBRTCClient.instance().removeDelegates does not remove delegates.

### ISSUE THREAD

**Issue by xpressmobtech** [QuickBlox/quickblox-ios-sdk#657]
*Help avoid duplicate issue reports, check [existing issues](https://github.com/QuickBlox/quickblox-ios-sdk/issues)*
 
**Environment details**
*iOS version, Quickblox iOS SDK version, QuickbloxWebRTC SDK version (optional)*
ios 10.1.1, QB SDK 2.8, QBRTC SDK 2.2

**Did this work before?**
No

**Expected behavior**
Should remove delegate when called.

**Actual behavior**
Doesn't actually remove delegate from list.

**Logs**
*Attach full logs that will describe your problem (please, use github [gist](https://gist.github.com/))*

rtc::<QBRTCClient 0x60000024ef70, delegates[2]> added <self.VideoChatViewController: 0x7ff0da493730> to list of delegates

after I popviewcontroller -

rtc::<QBRTCClient 0x60000024ef70, delegates[2]> removed <self.VideoChatViewController: 0x7ff0da51a860> from list of delegates

when I return to same viewcontroller -

rtc::<QBRTCClient 0x60000024ef70, delegates[3]> added <self.VideoChatViewController: 0x7ff0da4abcb0> to list of delegates

**Steps to reproduce the behavior**
add QBRTCClient delegate on viewDidLoad of page and remove delegate on viewDidDisappear, then return to page and see that no delegate was removed yet a new delegate is added.

**Any others comments?**
 
 this bug is causing all my QBRTC methods to fire for every delegate so if I leave my page and return all QBRTC methods fire twice, and if I leave and return 4 times all QBRTCClient delegates fire 4 times!

**Comment by xpressmobtech** [comment]
@Raikerian @Pro100Andrey @VitaliyGuru

**Comment by xpressmobtech** [comment]
@Raikerian @Pro100Andrey @VitaliyGuru any solutions for this issue?

**Comment by Pro100Andrey** [comment]
Hi @xpressmobtech,

rtc::<QBRTCClient 0x60000024ef70, delegates[2]> added <self.VideoChatViewController: 0x7ff0da493730> to list of delegates

after I popviewcontroller -

rtc::<QBRTCClient 0x60000024ef70, delegates[2]> removed <self.VideoChatViewController: 0x7ff0da51a860> from list of delegates

when I return to same viewcontroller -

rtc::<QBRTCClient 0x60000024ef70, delegates[3]> added <self.VideoChatViewController: 0x7ff0da4abcb0> to list of delegates

Answer :
0x7ff0da493730 != 0x7ff0da51a860 != 0x7ff0da4abcb0 

?

**Comment by Pro100Andrey** [comment]
@xpressmobtech QBRTCClient.instance addDelegate  - this NSHash​Table  (NSHashTableWeakMemory) storage 

> NSHashTableWeakMemory: Equal to NSPointerFunctionsWeakMemory. Uses weak read and write barriers. Using NSPointerFunctionsWeakMemory, object references will turn to NULL on last release.

**Comment by xpressmobtech** [comment]
I'm still new to iOS development, could you explain weak read & write barriers? How do I use them?

**Comment by xpressmobtech** [comment]
Just to clarify...In my viewdidload of viewcontroller1 I call QBRTCClient.instance().addDelegate(self) and on viewdiddisappear of viewcontroller1 I call QBRTCClient.instance().removeDelegate(self)...delegates should be at zero when my view did disappear however there Is  still one delegate and if I return to viewcontroller1 there is now 2 delegates because the delegates are never removed...how can I get these delegates removed after view did disappear?

**Comment by xpressmobtech** [comment]
@Pro100Andrey @VitaliyGuru @Raikerian  What do i do too solve my problem?

**Comment by VitaliyGuru** [comment]
@xpressmobtech , From your logs:
`rtc::<QBRTCClient 0x60000024ef70, delegates[2]> added <self.VideoChatViewController: 0x7ff0da493730> to the list of delegates`
So you add 0x7ff0da493730 to the list of delegate;

But 
`removed <self.VideoChatViewController: 0x7ff0da51a860> from list of delegates`

You removed 0x7ff0da51a860, not the same address of the instance that you added.

**Comment by xpressmobtech** [comment]
@VitaliyGuru @Pro100Andrey 
 logs:

viewDidLoad -- QBRTCClient.instance().addDelegate(self)
`rtc::<QBRTCClient 0x175cb900, delegates[1]> added <self.VideoChatViewController: 0x1777a7b0> to list of delegates_`

It says it added one viewcontroller "0x1777a7b0" to delegates list.

viewDidDisappear -- QBRTCClient.instance().removeDelegate(self)
`2016-11-17 08:05:45:298 self[1911:1180061] rtc::<QBRTCClient 0x175cb900, delegates[1]> removed <self.VideoChatViewController: 0x1777a7b0> from list of delegates`

Notice it says it removed the viewcontroller "0x1777a7b0" from list of  delegates, however as you can see there is still one delegate in the list, when there should be 0, which means it never removed anything, now if I instantiate the very same view once again, it adds another delegate to the list 

`rtc::<QBRTCClient 0x175cb900, delegates[2]> added <self.VideoChatViewController: 0x176b1320> to list of delegates`

Trying to remove delegate when viewDidDisappear.

`rtc::<QBRTCClient 0x175cb900, delegates[2]> removed <self.VideoChatViewController: 0x176b1320> from list of delegates`

it still has both the original delegate "0x1777a7b0" and the 2nd delegate added "0x176b1320", QBRTCClient.instance().addDelegate(self) adds delegates to QBRTCClient delegates list fine, but QBRTCClient.instance().removeDelegate(self) does not actually remove anything from the delegates list, causing there to be a delegate added for every page transition(which could in the 100s) , also all rtc methods are fired one time for every one delegate...which is a huge issue.

ANY SOLUTIONS ON HOW TO REMOVE THE DELEGATES FROM THE LIST SO THAT THE LIST IS AT ZERO?

**Comment by xpressmobtech** [comment]
and still no header files for QUICKBLOx 2.9 update https://github.com/QuickBlox/quickblox-ios-sdk/issues/659

**Comment by xpressmobtech** [comment]
@Raikerian  @VitaliyGuru @Pro100Andrey any solutions?

**Comment by xpressmobtech** [comment]
@Raikerian @VitaliyGuru @Pro100Andrey any solutions?

**Comment by Pro100Andrey** [comment]
Hi @xpressmobtech, please check:
```objc
    QBMulticastDelegate *observer = [[QBRTCClient instance] valueForKey:@"observer"];
    NSLog(@"before remove - %@", observer.delegates);
    [observer removeDelegate:self];
    NSLog(@"after remove %@", observer.delegates);
```

**Comment by xpressmobtech** [comment]
Works great, THANKS! Still don't know why QBRTCCLIENT.instance().removeDelegate() doesn't perform its function

**Comment by Pro100Andrey** [comment]
Hi @xpressmobtech, sorry this is bug from our side, we will be fix it of SDK 2.3

**Comment by Pro100Andrey** [comment]
@xpressmobtech Thanks !!!
